service/ui_plugin_build_baron.go

```diff
@@ -321,43 +386,270 @@ func (aes *altEndpointSuggest) parseResponse(r io.ReadCloser) ([]thirdparty.Jira
+	fallbackCtx := context.TODO()
+	fallbackChan := make(chan result, 1)
+	go func() {
+		suggestions, err := mss.jiraSuggester.Suggest(fallbackCtx, t)
+		fallbackChan <- result{suggestions, err}
+		close(fallbackChan)
+	}()
+
+	altEndpointTimeout := mss.altSuggester.GetTimeout()
+	altEndpointCtx, altEndpointCancel := context.WithTimeout(context.Background(), altEndpointTimeout)
+	defer altEndpointCancel()
+	suggestions, err := mss.altSuggester.Suggest(altEndpointCtx, t)
+
+	// If the alternative endpoint didn't respond quickly enough or didn't have results available,
+	// then we wait for the fallback results. Ideally we'd otherwise be able to cancel the request
+	// for fetching the fallback results, but we instead just return back to the caller without
+	// waiting for the associated goroutine to complete.
 	if err != nil {
-		return nil, err
+		grip.Warning(message.WrapError(err, message.Fields{
+			"message":   "failed to get results from alternative endpoint",
+			"task_id":   t.Id,
+			"execution": t.Execution,
+		}))
+
+		fallbackChanRes := <-fallbackChan
+		return fallbackChanRes.Tickets, JIRASource, fallbackChanRes.Error
 	}
 
-	if aes.bbProj.AlternativeEndpointUsername != "" {
-		req.SetBasicAuth(aes.bbProj.AlternativeEndpointUsername, aes.bbProj.AlternativeEndpointPassword)
+	return suggestions, BFSuggestionSource, nil
+}
+
+/////////////////////////////////
+// BF Suggestion Server Client //
+/////////////////////////////////
+
+func getBFSuggestionClient(bbProj evergreen.BuildBaronProject) (bfSuggestionClient, bool) {
+	var client bfSuggestionClient
+
+	if bbProj.BFSuggestionServer == "" {
+		return client, false
+	}
+	return bfSuggestionClient{
+		bbProj.BFSuggestionServer,
+		bbProj.BFSuggestionUsername,
+		bbProj.BFSuggestionPassword}, true
+}
+
+const suggestionPath = "/suggestions/{task_id}/{execution}"
+const getFeedbackPath = "/feedback/{task_id}/{execution}/user_id/{user_id}"
+const sendFeedbackPath = "/feedback/{task_id}/{execution}"
+const removeFeedbackPath = "/feedback/{task_id}/{execution}/user_id/{user_id}/feedback_type/{feedback_type}"
+
+type bfSuggestionClient struct {
+	server   string
+	username string
+	password string
+}
+
+// Send a request to the BF Suggestion server.
+func (bfsc *bfSuggestionClient) request(ctx context.Context, req *http.Request) (io.ReadCloser, error) {
+	client := util.GetHTTPClient()
+	defer util.PutHTTPClient(client)
+
+	if bfsc.username != "" {
+		req.SetBasicAuth(bfsc.username, bfsc.password)
 	}
 
 	req = req.WithContext(ctx)
 	resp, err := client.Do(req)
-	if resp != nil {
-		defer resp.Body.Close()
-	}
 	if err != nil {
 		return nil, err
 	}
-
 	if resp.StatusCode >= 300 || resp.StatusCode < 200 {
 		body, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
 			return nil, errors.Errorf("HTTP request returned unexpected status: %v", resp.Status)
 		}
 		return nil, errors.Errorf("HTTP request returned unexpected status=%v: %s", resp.Status, string(body))
 	}
+	return resp.Body, nil
```

Comment:
It's quite important that we close the body in the form/position of line 343 that you removed, and even though the body is closeable in its current form, as we'll leak resources in error cases.

Reply:
Updated the function to return a `[]byte` and close the body.